contracts/Paper.sol

```diff
@@ -0,0 +1,156 @@
+    uint256 public tokenIdEnd = 8000;
+
+    // Give out 50% of tokens, evenly split across each NFT
+    uint256 public supply = 1000000000 * (10**decimals());
+    uint256 public paperPerTokenId = supply / 2 / tokenIdEnd ;
+    
+    // track claimedTokens
+    mapping(uint256 => bool) public claimedByTokenId;
+    
+    // TODO verify?
+    address public daoAddress = 0x8707276DF042E89669d69A177d3DA7dC78bd8723;
+    
+    constructor() ERC20("Paper", "PAPER") {
+        transferOwnership(daoAddress);
+        dopeWarsLootContract = IERC721Enumerable(dopeWarsLootContractAddress);
+        _mint(daoAddress, supply / 2);
+    }
+
+    function snapshot() public onlyOwner {
+        _snapshot();
+    }
+
+    // The following functions are overrides required by Solidity.
+
+    function _beforeTokenTransfer(address from, address to, uint256 amount)
+        internal
+        override(ERC20, ERC20Snapshot)
+    {
+        super._beforeTokenTransfer(from, to, amount);
+    }
+
+    /// @notice Allows the DAO to set a new contract address for Dope Wars Loot. This is
+    /// relevant in the event that DWL migrates to a new contract.
+    /// @param dopeWarsLootContractAddress_ The new contract address for DWL
+    function daoSetLootContractAddress(address dopeWarsLootContractAddress_)
+        external
+        onlyOwner
+    {
+        dopeWarsLootContractAddress = dopeWarsLootContractAddress_;
+        dopeWarsLootContract = IERC721Enumerable(dopeWarsLootContractAddress);
+    }
+
+    /// @notice Claim Paper for a given Dope Wars Loot ID
+    /// @param tokenId The tokenId of the Dope Wars Loot NFT
+    function claimById(uint256 tokenId) external {
+        // Follow the Checks-Effects-Interactions pattern to prevent reentrancy
+        // attacks
+
+        // Checks
+
+        // Check that the msgSender owns the token that is being claimed
+        require(
+            _msgSender() == dopeWarsLootContract.ownerOf(tokenId),
+            "MUST_OWN_TOKEN_ID"
+        );
+
+        // Further Checks, Effects, and Interactions are contained within the
+        // _claim() function
+        _claim(tokenId, _msgSender());
+    }
+
+    /// @notice Claim Paper for all tokens owned by the sender
+    /// @notice This function will run out of gas if you have too much loot! If
+    /// this is a concern, you should use claimRangeForOwner and claim Dope in
+    /// batches.
+    function claimAllForOwner() external {
+        uint256 tokenBalanceOwner = dopeWarsLootContract.balanceOf(_msgSender());
+
+        // Checks
+        require(tokenBalanceOwner > 0, "NO_TOKENS_OWNED");
+
+        // i < tokenBalanceOwner because tokenBalanceOwner is 1-indexed
+        for (uint256 i = 0; i < tokenBalanceOwner; i++) {
+            // Further Checks, Effects, and Interactions are contained within
+            // the _claim() function
+            _claim(
+                dopeWarsLootContract.tokenOfOwnerByIndex(_msgSender(), i),
```

Comment:
This is inherited from the OZ contract ERC721Enumerable:

https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/token/ERC721/extensions/ERC721Enumerable.sol#L36